[C23] Implement N3018: The constexpr specifier for object definitions

Fznamznon opened the pull request:
The implementation mostly reuses C++ code paths where possible, including narrowing check in order to provide diagnostic messages in case initializer for constexpr variable is not exactly representable in target type.

The following won't work due to lack of support for other features:
- Diagnosing of underspecified declarations involving constexpr
- Constexpr attached to compound literals

Also due to lack of support for char8_t some of examples with utf-8 strings don't work properly.

Fixes https://github.com/llvm/llvm-project/issues/64742

Fznamznon committed:
[C23] Implement N3018: The constexpr specifier for object definitions

The implementation mostly reuses C++ code paths where possible,
including narrowing check in order to provide diagnostic messages in
case initializer for constexpr variable is not exactly representable in
target type.

The following won't work due to lack of support for other features:
- Diagnosing of underspecified declarations involving constexpr
- Constexpr attached to compound literals

Also due to lack of support for char8_t some of examples with utf-8
strings don't work properly.

Fznamznon commented:
cc @to268 , GitHub doesn't let me add you to the reviewers list.

tbaederr commented on clang/lib/AST/ExprConstant.cpp:
```diff
@@ -2231,6 +2231,13 @@ static bool CheckLValueConstantExpression(EvalInfo &Info, SourceLocation Loc,
     return false;
   }
 
+  if (Info.getLangOpts().C23) {
+    auto *VarD = dyn_cast_or_null<VarDecl>(BaseVD);
+    if (VarD && VarD->isConstexpr() && !LVal.isNullPointer()) {
+      Info.report(Loc, diag::err_c23_constexpr_pointer_not_null);
+    }
+  }
```
```suggestion
  if (Info.getLangOpts().C23) {
    if (const auto *VarD = dyn_cast_if_present<VarDecl>(BaseVD);
        VarD && VarD->isConstexpr() && !LVal.isNullPointer())
      Info.report(Loc, diag::err_c23_constexpr_pointer_not_null);
  }
```

tbaederr commented on clang/lib/AST/ExprConstant.cpp:
```diff
@@ -4110,6 +4117,10 @@ static CompleteObject findCompleteObject(EvalInfo &Info, const Expr *E,
 
     bool IsConstant = BaseType.isConstant(Info.Ctx);
+    bool ConstexprVar = false;
+    if (const auto *VD = dyn_cast_or_null<VarDecl>(
```
```suggestion
    if (const auto *VD = dyn_cast_if_present<VarDecl>(
```

tbaederr commented on clang/lib/Sema/SemaDecl.cpp:
```diff
@@ -8605,6 +8620,27 @@ static bool checkForConflictWithNonVisibleExternC(Sema &S, const T *ND,
+  }
+
+  if (T->isRecordType()) {
+    RecordDecl *RD = T->getAsRecordDecl();
```
```suggestion
    const RecordDecl *RD = T->getAsRecordDecl();
```

Fznamznon committed:
Fix format, apply comments

Fznamznon commented:
Ping.

AaronBallman commented on clang/lib/Sema/SemaDecl.cpp:
```diff
@@ -14240,6 +14294,114 @@ StmtResult Sema::ActOnCXXForRangeIdentifier(Scope *S, SourceLocation IdentLoc,
+}
+
+static bool checkC23ConstexprInitializer(Sema &S, const Expr *Init) {
+  const Expr *InitNoCast = Init->IgnoreImpCasts();
```
Probably want to ignore parens here as well; worth a test case:
```
constexpr int i = (12);
constexpr int j = (i);
```

Fznamznon replied:
Added test in `Sema/constexpr.c`.

AaronBallman commented on clang/lib/Sema/SemaDecl.cpp:
```diff
@@ -14240,6 +14294,114 @@ StmtResult Sema::ActOnCXXForRangeIdentifier(Scope *S, SourceLocation IdentLoc,
+    if (checkC23ConstexprInitConversion(S, Init))
+      return true;
+
+  if (auto *SE = dyn_cast<StringLiteral>(Init))
```
```suggestion
  if (const auto *SE = dyn_cast<StringLiteral>(Init))
```

AaronBallman commented on clang/lib/Sema/SemaDecl.cpp:
```diff
@@ -14240,6 +14294,114 @@ StmtResult Sema::ActOnCXXForRangeIdentifier(Scope *S, SourceLocation IdentLoc,
+      return true;
+
+  for (const Stmt *SubStmt : Init->children()) {
+    const Expr *ChildExpr = dyn_cast_or_null<const Expr>(SubStmt);
```
```suggestion
    const Expr *ChildExpr = dyn_cast_or_null<Expr>(SubStmt);
```

AaronBallman commented on clang/lib/Sema/SemaDecl.cpp:
```diff
@@ -14240,6 +14294,114 @@ StmtResult Sema::ActOnCXXForRangeIdentifier(Scope *S, SourceLocation IdentLoc,
+    if (!ChildExpr)
+      continue;
+
+    if (checkC23ConstexprInitializer(S, ChildExpr)) {
+      return true;
+    }
```
```suggestion
    if (checkC23ConstexprInitializer(S, ChildExpr))
      return true;
```

AaronBallman commented on clang/lib/Sema/SemaDecl.cpp:
```diff
@@ -14397,17 +14559,21 @@ void Sema::CheckCompleteVariableDeclaration(VarDecl *var) {
   QualType baseType = Context.getBaseElementType(type);
   bool HasConstInit = true;
 
+  if (getLangOpts().C23 && var->isConstexpr() && !Init)
+    Diag(var->getLocation(), diag::err_constexpr_var_requires_const_init)
+        << var;
```
Is there a reason the existing logic below doesn't suffice and we need a separate code path here?

Fznamznon replied:
Yes, existing logic uses `InitializationSequence` code paths that are deeply tied to C++ initialization rules, modifying these for C didn't seem reasonable.

AaronBallman commented on clang/test/C/C2x/n3018.c:
```diff
@@ -0,0 +1,86 @@
+  constexpr double onethird          = 1.0/3.0;
+  constexpr double onethirdtrunc     = (double)(1.0/3.0);
+
+  constexpr char string[] = { "\xFF", };
```
It'd be good to have a test with wide strings and another one with concatenated strings. e.g.,
```
constexpr char string[] = "test""ing this out\xFF";
```

Fznamznon replied:
Added in `test/Sema/constexpr.c`.

AaronBallman reviewed (commented):
Thank you for working on this! First round of comments done (I've not thoroughly reviewed the test cases yet though).

cor3ntin commented on clang/lib/Sema/SemaDecl.cpp:
```diff
@@ -8857,7 +8894,9 @@ void Sema::CheckVariableDeclarationType(VarDecl *NewVD) {
   if (NewVD->isConstexpr() && !T->isDependentType() &&
       RequireLiteralType(NewVD->getLocation(), T,
-                         diag::err_constexpr_var_non_literal)) {
+                         getLangOpts().C23
+                             ? diag::err_c23_constexpr_invalid_type
+                             : diag::err_constexpr_var_non_literal)) {
```
`RequireLiteralType` assumes C++ (ie it will cast to CXXRecordDecl), so maybe we need to duplicate that entirely.
Maybe just renaming `CheckC23ConstexprVarTypeQualifiers` to `CheckC23ConstexprVarType` and do the VLAs test there

cor3ntin commented on clang/lib/Sema/SemaDecl.cpp:
```diff
@@ -14240,6 +14294,114 @@ StmtResult Sema::ActOnCXXForRangeIdentifier(Scope *S, SourceLocation IdentLoc,
+  uint32_t BitWidth = SemaRef.Context.getTypeSize(CharType);
+  bool isUnsigned = CharType->isUnsignedIntegerType();
+  llvm::APSInt Value(BitWidth, isUnsigned);
+  const StringRef S = SE->getBytes();
+  for (unsigned I = 0, N = SE->getLength(); I != N; ++I) {
+    Value = S[I];
+    if (Value != S[I]) {
+      SemaRef.Diag(Loc, diag::err_c23_constexpr_init_not_representable)
+          << S[I] << CharType;
```
Nice catch. Yes, we should use `getCodeUnit` here - `SE->getLength()` already gives you the number of code units, something like 

```suggestion
  for (unsigned I = 0, N = SE->getLength(); I != N; ++I) {
   uint32_t C = SE->getCodeUnit(I); 
   Value = C;
    if (Value != C) {
      SemaRef.Diag(Loc, diag::err_c23_constexpr_init_not_representable)
          << C << CharType;
```

Fznamznon replied:
I'm not sure? What I'm trying to do is catch an overflowing conversion which already happens in `getCodeUnit()`. Perhaps if I have to use `getCodeUnit()` there should be another check then?

cor3ntin replied:
Oh wait.
How would you get into a situation where a string literal contains non representable data?
StringLiterals ought to be representable per constructions
https://eel.is/c++draft/lex.string#10.2.4
https://eel.is/c++draft/lex.string#10.1.sentence-2

Now that i understand what you are trying to do, I'm not sure that whole function is necessary at all.

Is there am example that concerns you?

Fznamznon replied:
What I'm trying to implement is  6.7.1 p5 of C (looking at N3096 draft):

> An object declared with storage-class specifier constexpr or any of its members, even recursively,
shall not have an atomic type, or a variably modified type, or a type that is volatile or restrict
qualified. An initializer of floating type shall be evaluated with the translation-time floating-point
environment. The declaration shall be a definition and shall have an initializer.139) **The value of
any constant expressions or of any character in a string literal of the initializer shall be exactly
representable in the corresponding target type**; no change of value shall be applied

There is also an illustrating example in p18:

>  Similarly, implementation-defined behavior related to the char type of the elements of the string literal "\xFF"
may cause constraint violations at translation time:
```
constexpr char string[] = { "\xFF", }; // ok
constexpr char8_t u8string[] = { u8"\xFF", }; // ok
constexpr unsigned char ucstring[] = { "\xFF", }; // possible constraint
// violation
```
> In both the string and ucstring initializers, the initializer is a (brace-enclosed) string literal of type char. If the type char is
capable of representing negative values and its width is 8, then the code above is equivalent to:
```
constexpr char string[] = { -1, 0, }; // ok
constexpr char8_t u8string[] = { 255, 0, }; // ok
constexpr unsigned char ucstring[] = { -1, 0, }; // constraint violation
```

About
> StringLiterals ought to be representable per constructions
https://eel.is/c++draft/lex.string#10.2.4
https://eel.is/c++draft/lex.string#10.1.sentence-2

My understanding that it is not the same for C. At least I fail to find somewhat same in C draft.

Fznamznon replied:
I would prefer adding a new function then, since returning an `int64` from `getCodeUnit` breaks several tests related to printing string literals in diagnostic and ast dumps.

cor3ntin replied:
Yes, because `wchar_t` can be signed.

Here is something that might work and avoid duplicating half of `StringLiteral` with dubious casts:
Get the unsigned value, and manually do the two complement to sign cast using `APInt`.

```
int64_t V = str.getCodeUnit(I); 
if(str.isOrdinary() || str.isWide()) {
    Width W = str.getCharByteWidth();
    APInt  apint(str.getCharByteWidth(), (uint64_t)V);
    apint.sext(Width + 1).trunc(Width);
    V = apint.getExtValue();
}
```

the `sext` idea credit of @erichkeane . I haven't tested it though.

Fznamznon committed:
Merge branch 'main' into N3018-1

Fznamznon commented on clang/include/clang/Basic/DiagnosticSemaKinds.td:
```diff
@@ -2932,6 +2932,22 @@ def warn_private_extern : Warning<
   "use __attribute__((visibility(\"hidden\"))) attribute instead">;
 
+// C23 constexpr
+def err_c23_thread_local_constexpr : Error<
+  "thread-local storage is not allowed with constexpr">;
+def err_c23_extern_constexpr : Error<
+  "extern specifier is not allowed with constexpr">;
```
I'm not quite sure reusing `err_invalid_decl_spec_combination` will look clear enough, here is what I got while trying:
```
t3.c:2:1: error: cannot combine with previous 'extern' declaration specifier
    2 | constexpr extern int V79 = 10;
      | ^         ~~~~~~

```
If that is still better, I can update it.

Fznamznon replied:
I ended up updating this.

Fznamznon commented on clang/lib/AST/ExprConstant.cpp:
```diff
@@ -4110,6 +4116,10 @@ static CompleteObject findCompleteObject(EvalInfo &Info, const Expr *E,
 
     bool IsConstant = BaseType.isConstant(Info.Ctx);
+    bool ConstexprVar = false;
+    if (const auto *VD = dyn_cast_if_present<VarDecl>(
+            Info.EvaluatingDecl.dyn_cast<const ValueDecl *>()))
+      ConstexprVar = VD->isConstexpr();
```
This is to check that an initializer for a `constexpr` variable in c23 can only reference other `constexpr` variables. Done since C++ allows things like
```
const int a = 0;
constexpr int b = a;
```

Fznamznon commented on clang/lib/Sema/SemaDecl.cpp:
```diff
@@ -14240,6 +14294,114 @@ StmtResult Sema::ActOnCXXForRangeIdentifier(Scope *S, SourceLocation IdentLoc,
                                                       : IdentLoc);
 }
 
+static ImplicitConversionKind getConversionKind(QualType FromType,
+                                                QualType ToType) {
+  if (ToType->isIntegerType()) {
+    if (FromType->isComplexType())
+      return ICK_Complex_Real;
+    if (FromType->isFloatingType())
+      return ICK_Floating_Integral;
+    if (FromType->isIntegerType())
+      return ICK_Integral_Conversion;
+  }
+
+  if (ToType->isFloatingType()) {
+    if (FromType->isComplexType())
+      return ICK_Complex_Real;
+    if (FromType->isFloatingType())
+      return ICK_Floating_Conversion;
+    if (FromType->isIntegerType())
+      return ICK_Floating_Integral;
+  }
+
+  return ICK_Identity;
+}
```
> I wonder if we should expose (some of) the logic of IsStandardConversion() from SemaOverload.cpp

You mean part starting from
https://github.com/llvm/llvm-project/blob/af65379e383bac651f0868237e9086630b15ee0d/clang/lib/Sema/SemaOverload.cpp#L2074 ?

If yes, how this will help better with the problem of adding new conversions?

Fznamznon commented on clang/lib/Sema/SemaDecl.cpp:
```diff
@@ -9237,6 +9282,22 @@ static FunctionDecl *CreateNewFunctionDecl(Sema &SemaRef, Declarator &D,
+      (SemaRef.getLangOpts().C23 &&
+       ConstexprKind == ConstexprSpecKind::Constexpr)) {
+
+    if (SemaRef.getLangOpts().C23)
+      SemaRef.Diag(D.getDeclSpec().getConstexprSpecLoc(),
+                   diag::err_c23_constexpr_not_variable);
```
> Hmmm, I'm not certain we need to change anything here

Well, without that move for C we silently create `FunctionDecl` and exit whether it has `constexpr` specifier or not.

> it would be more clear to say that a function cannot be constexpr in C instead of saying you declared something constexpr that isn't a variable. WDYT?

I don't seem to have a problem with this particular message, for C++ similar invalid use cases clang says and does the same. But If that will help users, ok.
I also use `err_c23_constexpr_not_variable` message when diagnosing constexpr applied to a type declaration. Should I add more specific message for this case as well?

Fznamznon committed:
Apply some of the feedback:

- Minor changes
- Don't reuse requireLiteralType
- Diagnose interaction with extern and thread_local from
  DeclSpec::Finish()
- Add a couple of test cases

Fznamznon committed:
Update a release note

Fznamznon committed:
Remove unused messages

Fznamznon committed:
Merge branch 'main' into N3018-1

Fznamznon committed:
Merge branch 'main' into N3018-1

Fznamznon committed:
Update comments

Fznamznon committed:
Merge branch 'main' into N3018-1

Fznamznon committed:
Use getCodeUnit

cor3ntin commented on clang/include/clang/AST/Expr.h:
```diff
@@ -1914,6 +1914,17 @@ class StringLiteral final
   }
 
+  // Get code unit but preserve sign info.
+  int64_t getCodeUnitS(size_t I, uint64_t ByteWidth) const {
```
```suggestion
  int64_t getCodeUnitS(size_t I, uint64_t BitWidth) const {
```
ByteWidth is actually a number of bits, right?

Fznamznon replied:
It is a number of bits in byte, but ok

cor3ntin commented on clang/lib/Sema/SemaDecl.cpp:
```diff
@@ -13816,7 +13862,9 @@ void Sema::AddInitializerToDecl(Decl *RealDecl, Expr *Init, bool DirectInit) {
 
     // C99 6.7.8p4. All file scoped initializers need to be constant.
-    if (!getLangOpts().CPlusPlus && !VDecl->isInvalidDecl())
+    // Avoid double diagnosing for constexpr variables.
```
```suggestion
    // Avoid double diagnoses for constexpr variables.
```

cor3ntin commented on clang/lib/Sema/SemaDecl.cpp:
```diff
@@ -14270,6 +14318,113 @@ StmtResult Sema::ActOnCXXForRangeIdentifier(Scope *S, SourceLocation IdentLoc,
+    int64_t C = SE->getCodeUnitS(I, SemaRef.Context.getCharWidth());
+    Value = C;
+    if (Value != C) {
+      SemaRef.Diag(Loc, diag::err_c23_constexpr_init_not_representable)
```
Maybe we can use `getLocationOfByte` instead of `Loc` here, if it's easy (I'm not insisting)

cor3ntin reviewed (commented):
Modulo comment I'm happy with the StringLiteral aspect of the patch and generally the rest of the patch looks good to me, but I'm not familiar enough with the peculiarities of C to provide an in-depth review

Fznamznon committed:
Merge branch 'main' into N3018-1

Fznamznon committed:
Apply suggestions

to268 commented on clang/test/Sema/constexpr.c:
```diff
@@ -0,0 +1,287 @@
+constexpr static auto V5 = 1;
+constexpr static const auto V6 = 1;
+constexpr static const int V7 = 1;
+constexpr static int V8 = 1;
```
I would preferably suggest to add a couple of `constexpr auto` tests, theses cases bellow should be enough to make sure that no divergence has been introduced.
```c
constexpr auto Ulong = 1L;
constexpr auto CompoundLiterral = (int){13};
constexpr auto DoubleCast = (double)(1 / 3);
constexpr auto String = "this is a string";
constexpr signed auto Long = 1L; // expected-error {{'auto' cannot be signed or unsigned}}
```

Adding some `_Static_assert` would be better to also make sure that the `auto` keyword is on par when adding the `constexpr` keyword.
```c
_Static_assert(_Generic(Ulong, long : 1));
_Static_assert(_Generic(CompoundLiterral, int : 1));
_Static_assert(_Generic(DoubleCast, double : 1));
_Static_assert(_Generic(String, char* : 1));
```

Fznamznon replied:
Makes sense, thanks, done.

to268 commented:
I have found no major issues with the C23 `auto` keyword.
There is a minor improvement that I've found in the test cases.

Concerning N3006 (AKA: Underspecified object declarations), I haven't reviewed yet the compatibility with the `constexpr` keyword, since I haven't work much on it yet.

Fznamznon committed:
Merge branch 'main' into N3018-1

Fznamznon committed:
Add test improvement

Fznamznon commented:
Adding more reviewers since Aaron is on vacation...

Fznamznon commented:
Ping.

Fznamznon commented:
Ping.

Fznamznon committed:
Merge branch 'main' into N3018-1

Fznamznon committed:
Target clang 19

Fznamznon commented:
Ping.

AaronBallman commented on clang/test/Parser/c23-constexpr.c:
```diff
@@ -0,0 +1,6 @@
+// RUN: %clang_cc1 -fsyntax-only -verify=c23 -std=c23 %s -Wpre-c2x-compat
+// RUN: %clang_cc1 -fsyntax-only -verify=c17 -std=c17 %s
+
+
```
```suggestion

```

AaronBallman replied:
We should probably add some additional parsing tests here that show we properly parse C oddities like:
```
int static constexpr i = 1; // Accepted
void func(int array[constexpr]); // Rejected (invalid syntax)
_Atomic constexpr int i = 0; // Rejected (invalid combo)
```

Fznamznon replied:
Added test cases.

AaronBallman commented on clang/test/Sema/constexpr.c:
```diff
@@ -0,0 +1,296 @@
+constexpr void f0() {} // expected-error {{'constexpr' can only be used in variable declarations}}
+constexpr const int f1() { return 0; } // expected-error {{'constexpr' can only be used in variable declarations}}
+
+constexpr struct S1 { int f; }; //expected-error {{struct cannot be marked constexpr}}
```
This will become an underspecified object declaration; CC @to268 to make sure he's covering this with his patch. Same is true for lines 8-11 as well.

AaronBallman commented on clang/test/Sema/constexpr.c:
```diff
@@ -0,0 +1,296 @@
+constexpr auto Ulong = 1L;
+constexpr auto CompoundLiteral = (int){13};
+constexpr auto DoubleCast = (double)(1 / 3);
+constexpr auto String = "this is a string";
```
This should be diagnosed as an error because it's a pointer declaration and the initializer is not a null pointer value.

Fznamznon replied:
Fixed now.

AaronBallman commented on clang/test/Sema/constexpr.c:
```diff
@@ -0,0 +1,296 @@
+
+constexpr thread_local int V11 = 38; // expected-error {{cannot combine with previous '_Thread_local' declaration specifier}}
+constexpr static thread_local double V12 = 38; // expected-error {{cannot combine with previous '_Thread_local' declaration specifier}}
+constexpr extern thread_local char V13; // expected-error {{cannot combine with previous '_Thread_local' declaration specifier}}
```
Good test case to add:
```
constexpr extern char Oops = 1; // Can't combine constexpr with extern
```

Fznamznon replied:
Added, line 70 in the same file.

AaronBallman commented on clang/test/Sema/constexpr.c:
```diff
@@ -0,0 +1,296 @@
+
+constexpr volatile int V17 = 0; // expected-error {{constexpr variable cannot have type 'const volatile int'}}
+
+constexpr int * restrict V18 = 0; // expected-error {{constexpr variable cannot have type 'int *const restrict'}}
```
Another test case to consider adding: `constexpr int * restrict * Foo = 0; // Accepted`

Fznamznon replied:
Added, line 71.

AaronBallman commented on clang/test/Sema/constexpr.c:
```diff
@@ -0,0 +1,296 @@
+constexpr int *V85 = nullptr;
+
+// Check that constexpr variables should not be VLAs.
+void f6(const int P1) {
```
We should also check for VM types:
```
void func(int n, int array[n]) {
  constexpr typeof(array) foo = 0; // Should be accepted because array is a pointer type, not a VLA type

  int (*(*fp)(int a))[a];
  constexpr typeof(fp) bar = 0; // Should be rejected because `fp` is a variably modified type but not a VLA

}
```

Fznamznon replied:
Yep, checked at the end of the test.

AaronBallman reviewed (commented):
Sorry for the delay in review on this! I think it's coming along well, though I did spot some concerns or additional test coverage.

Fznamznon committed:
Merge branch 'main' into N3018-1

Fznamznon committed:
Merge branch 'main' into N3018-1

Fznamznon committed:
Evaluate initializers in C++ fashion only for C23 constexpr vars

Caused test fail in test/Sema/atomic-expr.c with the new evaluator.

Fznamznon commented on clang/lib/Sema/SemaDecl.cpp:
```diff
@@ -14277,6 +14325,113 @@ StmtResult Sema::ActOnCXXForRangeIdentifier(Scope *S, SourceLocation IdentLoc,
                                                       : IdentLoc);
 }
 
+static ImplicitConversionKind GetConversionKind(QualType FromType,
```
Thanks for the review?
Just to double check that I understand right, are you suggesting to drop the whole approach that reuses C++ narrowing check due to need to keep this `ImplicitConversionKind` construction code, or you're suggesting to somehow find existing code that builds up `StandardConversionSequence` and reuse so there is no additional code that depends on these conversion kinds?

I think I can somehow integrate into `InitializationSequence`, but at the moment I'm having troubles with creating `StandardConversionSequence` object like how it is done there for C++ narrowing check because it has special initialization step which is used only for C++.

Fznamznon replied:
@AaronBallman , I think I could reuse `IsStandardConversion` to build `StandardConversionSequence` object and call it from `InitializationSequence::Perform`, so `InitializationSequence` will take care of the recursion.

AaronBallman replied:
That's along the lines of what I was thinking, yes. I was hoping we could use something like `Sema::TryImplicitConversion()` to get us an implicit conversion sequence. From there, we could check to see if it's a standard conversion sequence to get the object which gives access to `StandardConversionSequence::getNarrowingKind()` to see if it's been narrowed.

Fznamznon replied:
Ok, I've made an attempt to implement that.

Fznamznon committed:
Apply review feedback

- Move initializer checking to SemaInit
- Use TryImplicitSequence
- Fix null pointer initializer check, it was all wrong
- When doing type check, check canonical type
- Add test cases

Fznamznon committed:
Merge branch 'main' into N3018-1

Fznamznon commented on clang/lib/Sema/SemaDecl.cpp:
```diff
@@ -8569,6 +8571,37 @@ static bool checkForConflictWithNonVisibleExternC(Sema &S, const T *ND,
+static bool CheckC23ConstexprVarType(Sema &SemaRef, SourceLocation VarLoc,
+                                     QualType T) {
+
+  if (T->isVariableArrayType()) {
```
Uses `isVariablyModifiedType()`.

Fznamznon commented on clang/lib/Sema/SemaDecl.cpp:
```diff
@@ -8569,6 +8571,37 @@ static bool checkForConflictWithNonVisibleExternC(Sema &S, const T *ND,
+  // works on non-arrays as well).
+  T = SemaRef.Context.getBaseElementType(T);
+
+  // C23 6.7.1p4: An object declared with storage-class specifier constexpr or
```
Yup, done.

Fznamznon commented on clang/lib/Sema/SemaDecl.cpp:
```diff
@@ -5140,6 +5140,8 @@ Decl *Sema::ParsedFreeStandingDeclSpec(Scope *S, AccessSpecifier AS,
           << GetDiagnosticTypeSpecifierID(DS)
           << static_cast<int>(DS.getConstexprSpecifier());
+    else if (getLangOpts().C23)
+      Diag(DS.getConstexprSpecLoc(), diag::err_c23_constexpr_not_variable);
```
Ignoring per https://github.com/llvm/llvm-project/pull/73099#discussion_r1491320091 .

Fznamznon commented on clang/lib/Sema/SemaDecl.cpp:
```diff
@@ -13823,7 +13869,9 @@ void Sema::AddInitializerToDecl(Decl *RealDecl, Expr *Init, bool DirectInit) {
 
     // C99 6.7.8p4. All file scoped initializers need to be constant.
-    if (!getLangOpts().CPlusPlus && !VDecl->isInvalidDecl())
+    // Avoid double diagnoses for constexpr variables.
```
Done.

Fznamznon commented on clang/lib/Sema/SemaDecl.cpp:
```diff
@@ -8569,6 +8571,37 @@ static bool checkForConflictWithNonVisibleExternC(Sema &S, const T *ND,
+          return CheckC23ConstexprVarType(SemaRef, VarLoc, F->getType());
+        }))
+      return true;
+  }
```
Added. Lines 130-137 in test/Sema/constexpr.c.

Fznamznon commented on clang/include/clang/AST/Expr.h:
```diff
@@ -1862,6 +1862,17 @@ class StringLiteral final
   }
 
+  // Get code unit but preserve sign info.
+  int64_t getCodeUnitS(size_t I, uint64_t BitWidth) const {
```
FYI https://github.com/llvm/llvm-project/pull/73099#pullrequestreview-1765410034 .

Fznamznon committed:
Minors to the tests

Fznamznon commented on clang/test/C/C2x/n3018.c:
```diff
@@ -0,0 +1,86 @@
+// RUN: %clang_cc1 -std=c2x -verify -triple x86_64 -pedantic -Wno-conversion -Wno-constant-conversion %s
```
Done.

AaronBallman reviewed (commented):
I've done another pass over the changes and in general, I think we're really close to done. I spotted these two issues:
```
#define NAN __builtin_nan("1")
#define FLT_SNAN __builtin_nansf("1")

constexpr float f2 = (long double)NAN; // 6.7.2p21 EXAMPLE 5: should be accepted
constexpr double d3 = FLT_SNAN; // 6.7.2p21 EXAMPLE 5: should be rejected
```
but would be okay with handling those in a follow-up if they turned out to be complicated. Please let me know your thoughts.

Fznamznon committed:
Merge branch 'main' into N3018-1

Fznamznon committed:
Merge branch 'main' into N3018-1

Fznamznon committed:
Implement NaN cases

Fznamznon commented:
> Please let me know your thoughts.

@AaronBallman , implemented in https://github.com/llvm/llvm-project/pull/73099/commits/dff61e0f49d0ba0d868f4e9668f75718b031e1e3 .
I was mostly referring to the original page https://www.open-std.org/jtc1/sc22/wg14/www/docs/n3018.htm when implementing the feature, so these NaN cases slipped.

AaronBallman commented on clang/lib/Sema/SemaOverload.cpp:
```diff
@@ -414,20 +415,42 @@ NarrowingKind StandardConversionSequence::getNarrowingKind(
+        Converted.convert(Ctx.getFloatTypeSemantics(FromType),
+                          llvm::APFloat::rmNearestTiesToEven, &ignored);
+        if (Ctx.getLangOpts().C23) {
+          if (FloatVal.isNaN() && Converted.isNaN()) {
+            if (!FloatVal.isSignaling() && !Converted.isSignaling()) {
+              // Quiet NaNs are considered the same value, regardless of
+              // payloads.
+              return NK_Not_Narrowing;
+            }
+          }
```
```suggestion
          if (FloatVal.isNaN() && Converted.isNaN() &&
             !FloatVal.isSignaling() && !Converted.isSignaling()) {
            // Quiet NaNs are considered the same value, regardless of
            // payloads.
            return NK_Not_Narrowing;
          }
```
Combining some logic, NFC

Fznamznon replied:
done.

AaronBallman reviewed (approved):
LGTM aside from a simplification that you can fix up when landing. Thank you for this effort, this is fantastic work!

Fznamznon committed:
Apply a minor
